lib/THC/THCReduceApplyUtils.cuh

```diff
@@ -72,6 +72,76 @@ __device__ T reduceBlock(T* smem,
   return r;
 }
 
+// Return value is in threadVals for threadIdx.x == 0
+template <typename T, typename ReduceOp, int N>
```

Comment:
can you make `reduceBlock` call this function with `N == 1`?

there's a lot of duplicated code in this file.
Similarly, `reduceNBlock` versus `reduceBlockN` is really confusing naming. Perhaps this version should be `reduceMultipleBlock`should be this version, whereas the multiple values per thread could be `reduceBlock`, named the same as the N == 1 case, since it's reducing all of them.